### PR TITLE
[WIP] Refactor runv

### PR DESCRIPTION
**Pull request by gnawux**

### CODE REVIEW

**Comment by HuKeping** [comment]
seems conflict

**Comment by gnawux** [comment]
@HuKeping yes. just a preview, will rebase+squash during this weekend, together with adding back the associate.

File: hypervisor/lazy.go [hunk 1]
```diff
@@ -1,126 +1,2 @@
 package hypervisor
```
Comment (WeiZhang555) [inline hypervisor/lazy.go]: Everything is gone, why keep this line and this file ? :100:

Comment (gnawux) [inline hypervisor/lazy.go]: Just make rsync easier, will remove the empty files before merge

File: hypervisor/vm.go [hunk 2]
```diff
@@ -43,85 +46,65 @@ func (vm *Vm) ReleaseResponseChan(ch chan *types.VmResponse) {
...
-		if !ok || Response == nil || Response.Code == types.E_VM_SHUTDOWN {
-			vm.ReleaseResponseChan(Status)
-			vm.clients = nil
-			stop = true
-		}
-		if Response != nil {
-			glog.V(1).Infof("Got response: %d: %s", Response.Code, Response.Cause)
-		} else {
-			glog.V(1).Infof("Nil response from status chan")
+		Status <- &types.VmResponse{
```
Comment (gao-feng) [inline hypervisor/vm.go]: No need to return response to status, status was not assigned to anyone yet.

File: hypervisor/vm.go [hunk 3]
```diff
@@ -158,171 +132,226 @@ func (vm *Vm) ReleaseVm() (int, error) {
...
+	vm.ctx.SetNetworkEnvironment(config)
+	Status, err := vm.GetResponseChan()
+	if err != nil {
+		vm.ctx.Log(ERROR, "failed to get status chan to monitor startpod: %v", err)
+		return api.NewResultBase(vm.Id, false, err.Error())
 	}
+	defer vm.ReleaseResponseChan(Status)
 
-	vm.Hub <- runPodEvent
+	vm.ctx.startPod()
```
Comment (gao-feng) [inline hypervisor/vm.go]: Call ctx.startPod directly here, If there is no need to handle runpod Event in state machine?

File: hypervisor/context.go [hunk 4]
```diff
@@ -106,18 +96,28 @@ func InitContext(id string, hub chan VmEvent, client chan *types.VmResponse, dc
...
-		devices:         newDeviceMap(),
-		progress:        newProcessingList(),
 		lock:            &sync.Mutex{},
-		wait:            false,
-	}, nil
+	}
+	ctx.networks.sandbox = ctx
+
+	if dc == nil {
+		dc = HDriver.InitContext(homeDir)
```
Comment (gao-feng) [inline hypervisor/context.go]: The initialized dc has not been assigned to context.

File: hypervisor/network.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+}
+
+func (nc *NetworkContext) sandboxInfo() *hyperstartapi.Pod {
+
+	vmSpec := NewVmSpec()
+
+	vmSpec.Hostname = nc.Hostname
+	vmSpec.Dns = nc.Dns
+	vmSpec.PortmappingWhiteLists = &hyperstartapi.PortmappingWhiteList{
+		InternalNetworks: nc.Neighbors.InternalNetworks,
```
Comment (gao-feng) [inline hypervisor/network.go]: nc.Neighbors may be nil

File: hypervisor/container.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+func (cc *ContainerContext) configProcess() {
+	c := cc.ContainerDescription
+
+	envs := []hyperstartapi.EnvironmentVar{}
+	for e, v := range c.Envs {
+		envs = append(envs, hyperstartapi.EnvironmentVar{Env: e, Value: v})
+	}
+	cc.process = &hyperstartapi.Process{
+		User:             c.UGI.User,
```
Comment (gao-feng) [inline hypervisor/container.go]: UGI may be nil

**Comment by gnawux** [comment]
[WIP] rebase log: 

- rebased to current master (#385), 
- contents from some pr is missing, will re-enable them: #285, #286, #329, #351

**Comment by laijs** [comment]
failed on: The command "hack/verify-gofmt.sh" failed and exited with 1 during .

**Comment by gnawux** [comment]
@laijs I am working on fix the build and will push again later today

**Pull request closed without merging**